### ISSUE TITLE
Lack of or unclear or out of date debugging instructions in `yo office` template

### ISSUE THREAD

**Issue by tj-aew** [OfficeDev/office-js-docs-pr#3355]
# Provide required information needed to triage your issue

Having bootstrapped a new custom functions project via [`yo office`](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/yeoman-generator-overview), there don't seem to be clear debug steps (or the steps there don't work) and sadly I haven't been smart enough to get it working. The `README.md` says:

> This template supports debugging custom functions from [Visual Studio Code](https://code.visualstudio.com/). For more information see [Custom functions debugging](https://aka.ms/custom-functions-debug). For general information on debugging task panes and other Office Add-in parts, see [Test and debug Office Add-ins](https://docs.microsoft.com/office/dev/add-ins/testing/test-debug-office-add-ins).

That [Custom functions debugging](https://aka.ms/custom-functions-debug) link says it's purely for UI-less custom functions (whereas the generator includes a button and taskpane). Links from there take you to https://docs.microsoft.com/en-us/office/dev/add-ins/testing/test-debug-office-add-ins and https://docs.microsoft.com/en-us/office/dev/add-ins/testing/debug-add-ins-overview, but at that point I didn't see any tie-back to the template project and the debugging support it seems to set up.

I did try the steps in [Custom functions debugging](https://aka.ms/custom-functions-debug) in case they applied, but breakpoints weren't hit.

So I tried (I may be doing unnecessary steps):

1. `npm run build:dev` (and/or `npm run watch` or the equivalent via Terminal > Run Task...)
2. `npm run dev-server` (or equivalent through Terminal > Run Task... > Dev Server)
3. In VSCode, set a breakpoint
4. In VSCode, start debugging by going to Run and Debug and running "Excel Desktop (Custom Functions)"
5. `npm run start:desktop`

But again, breakpoints don't get hit.

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: PC desktop and Office on the web
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: 2203 Build 16.0.15028.20152
* Operating System: Windows 10
* Browser (if using Office on the web): Edge (Chromium)
 
## Expected behavior

I was expecting `README.md` to have specific steps for debugging, or to be able to figure it out from the `scripts` in `package.json`. (The `README.md` could really use a section explaining what each of them does, and why [for instance] `start` does `npm run build` but neither `start:desktop` nor `start:web` does. I *think* it's because you're expected to be running `watch` and `dev-server` in the background already, but...)

## Current behavior

The `README.md` seems to either not have/link to this information or the links are out of date or similar.
I've tried various different combinations of things but either the add-in loads but won't stop on breakpoints, or doesn't load.

## Steps to reproduce

(See above)

# Provide additional details

The project is the result of using the latest `yo office` to generate a really basic project:

```text
C:\temp>yo office

     _-----_     ?--------------------------?
    |       |    ¦   Welcome to the Office  ¦
    |--(o)--|    ¦   Add-in generator, by   ¦
   `---------´   ¦ @OfficeDev! Let's create ¦
    ( _´U`_ )    ¦    a project together!   ¦
    /___A___\   /?--------------------------?
     |  ~  |
   __'.___.'__
 ´   `  |° ´ Y `

? Choose a project type: Excel Custom Functions Add-in project
? Choose a script type: JavaScript
? What do you want to name your add-in? basic-functions
```

## Context

We've been struggling with debugging custom functions off and on for a while. This is an attempt to understand how it's supposed to work so we can get on with the actual work of writing the functions. :-) Thanks in advance!

**Comment by AlexJerabek** [comment]
@grangeryy, could you please also assist here?

**Comment by osmanm786** [comment]
anything on this UI-less custom functions debugging?

**Comment by osmanm786** [comment]
anything on this UI-less custom functions debugging?

**Comment by alison-mk** [comment]
Hi @tj-aew, thank you for bringing this to our attention and providing so much information. You're correct, we don't currently have an article that describes how to debug the default custom functions add-in installed by Yo Office. I'm researching this particular debugging scenario internally and will leave this GitHub issue open as I do that research. 

Is your add-in development blocked in any specific way? We would like to help unblock you if there are any particular errors that you're encountering. If not, thanks again for submitting this issue and we'll post more updates here as we gather information. 

Cheers, 
Alison

**Comment by tj-aew** [comment]
@alison-mk Thanks Alison. It's not currently getting in the way (though it did cost a lot of time), because I had an old custom functions project that does mostly work for debugging, so I used that instead (which sounds simple but was more complicated than that).

I'm not 100% sure it's only a documentation problem as @Rick-Kirkham said, though I agree it's reasonable to start there. Starting `Dev Server`, `Watch`, then (after waiting for builds to complete) doing `Debug: Excel Desktop` and running the debug task VS Code does show that it connects when Excel loads the functions, but breakpoints just don't work. (I also tried without `Watch`.) My suspicion is that there's an actual issue with the project the generator creates. Would love to know if either of you can get breakpoints to work with a freshly-scaffolded project, and if so what sequence you followed to do that (it doesn't need to be polished :-) ).

Thanks!

**Comment by alison-mk** [comment]
Hi @tj-aew, thanks again for reporting this issue. You're correct that there was a bug with the Excel Custom Functions project generator resulting in breakpoints not working during debugging.

Our engineering team located the bug with the project generator and deployed a fix. The fix for the TypeScript version of the Custom Functions project generator has been published. The fix for the JavaScript version will be published soon. I'll leave this issue open until both fixes are published. 

Alison

**Comment by cfis** [comment]
@alison-mk - Where was the fix published so we can look at it?

I have the same problem as above. If you follow these instructions:

https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-debugging

Debugging of custom functions does not work. Although the debugger does attach, all breakpoints are grayed out and have a message in the breakpoint window of "Unbound breakpoint"

**Comment by alison-mk** [comment]
@igor-ribeiiro Can you take a look at the comment posted yesterday by @cfis? 

> Where was the fix published so we can look at it?

> Debugging of custom functions does not work. Although the debugger does attach, all breakpoints are grayed out and have a message in the breakpoint window of "Unbound breakpoint"

**Comment by igor-ribeiiro** [comment]
> @alison-mk - Where was the fix published so we can look at it?
> 
> I have the same problem as above. If you follow these instructions:
> 
> https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-debugging
> 
> Debugging of custom functions does not work. Although the debugger does attach, all breakpoints are grayed out and have a message in the breakpoint window of "Unbound breakpoint"

Are you trying to debug on the web or desktop? The fix was for Excel Desktop.
You can get the latest version with: https://docs.microsoft.com/en-us/office/dev/add-ins/develop/yeoman-generator-overview in the Custom Functions option.
Could you try a new project? You might have an outdated version.

**Comment by cfis** [comment]
Yes, Excel Desktop, Windows 11. 

I believe I am running the latest version:

```
C:\Source>npm list -g
C:\Users\cfis\AppData\Roaming\npm
+-- eslint@8.16.0
+-- generator-office@1.8.7
`-- yo@4.3.0
```

From the link:

https://www.npmjs.com/package/generator-office

Looks like 1.8.7 from 25 days ago is the latest?

**Comment by tj-aew** [comment]
These instructions are still incorrect or incomplete nearly five months later. (And/or the debugging setup of the `yo office` scaffolded project are not working correctly.)

Also, despite what it says [on the debugging page](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-debugging) linking to this issue, it's not just JavaScript projects that are affected, TypeScript projects are as well.

I'd really like to be able to reliably debug custom functions. :-)

Steps to reproduce issues with the latest `yo office` generator:

1. Scaffold a new TypeScript custom functions add-in
2. Open [the debugging page](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-debugging)
3. Open the project in VS Code

At this point, the first thing you'll notice is that the project auto-runs the `Dev Server` task. But the instructions say:

> **Run your add-in from VS Code**
>
> 1. Open your custom functions root project folder in VS Code.
> 2. Choose `Terminal > Run Task `and type or select `Watch`. This will monitor and rebuild for any file changes.
> 3. Choose `Terminal > Run Task` and type or select `Dev Server`.

That would seem to suggest you should start `Watch` before `Dev Server`. Certainly the instructions should at least tell you that `Dev Server` may already be running.

Separately, the `Dev Server` task starts up and reaches this point where it says it's serving files:

![image](https://user-images.githubusercontent.com/101200432/186436841-cc382e1e-028c-4300-bbf6-43844665cd85.png)

So it seems like all is good and you can run `Watch` now. But actually, **it's still doing something**. Specifically, it's compiling the files. 32 seconds later (on my reasonable-spec but not gonzo system), we get:

![image](https://user-images.githubusercontent.com/101200432/186437013-98aa050b-0ef4-460c-973c-327b8236212e.png)

But okay, we've waited for that to be ready. Let's continue with my steps to reproduce the problem (not the quoted list above):

4. Run `Watch`.

**Sometimes**, that successfully compiles the project. Other times, you get an endless loop of failures:

![image](https://user-images.githubusercontent.com/101200432/186437524-bd50380f-cf1f-497e-abd3-1b6a0fa34a30.png)

```text
ERROR in C:\temp\new-custom\src\commands\commands.ts
./src/commands/commands.ts 46:33-36
[tsl] ERROR in C:\temp\new-custom\src\commands\commands.ts(46,34)
      TS2304: Cannot find name 'add'.

ERROR in C:\temp\new-custom\src\commands\commands.ts
./src/commands/commands.ts 47:35-40
[tsl] ERROR in C:\temp\new-custom\src\commands\commands.ts(47,36)
      TS2304: Cannot find name 'clock'.

ERROR in C:\temp\new-custom\src\commands\commands.ts
./src/commands/commands.ts 48:39-48
[tsl] ERROR in C:\temp\new-custom\src\commands\commands.ts(48,40)
      TS2304: Cannot find name 'increment'.

ERROR in C:\temp\new-custom\src\commands\commands.ts
./src/commands/commands.ts 49:33-43
[tsl] ERROR in C:\temp\new-custom\src\commands\commands.ts(49,34)
      TS2552: Cannot find name 'logMessage'. Did you mean 'onmessage'?

ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts
./src/taskpane/taskpane.ts 37:33-36
[tsl] ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts(37,34)
      TS2304: Cannot find name 'add'.

ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts
./src/taskpane/taskpane.ts 38:35-40
[tsl] ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts(38,36)
      TS2304: Cannot find name 'clock'.

ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts
./src/taskpane/taskpane.ts 39:39-48
[tsl] ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts(39,40)
      TS2304: Cannot find name 'increment'.

ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts
./src/taskpane/taskpane.ts 40:33-43
[tsl] ERROR in C:\temp\new-custom\src\taskpane\taskpane.ts(40,34)
      TS2552: Cannot find name 'logMessage'. Did you mean 'onmessage'?

8 errors have detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.52.1 compiled with 8 errors in 150 ms
```

I used to think the problem was `Watch` and `Dev Server` overlapping, but I've carefully waited for `Dev Server` to be done before doing `Watch`, and still had it happen. Also, since the instructions say to run `Watch` before `Dev Server`, I've also disabled the auto-run, run `Watch`, waited for the first compile, and then run `Dev Server`, and still had it happen.

Which raises the point: Why **two** of them? Wouldn't it make more sense to have **one** task that you need running for development, not two?? Especially as they both try to compile the code, and apparently sometimes conflict with each other?

Let's assume `Watch` didn't flip out and now each of them has completed a compile, and continue with the repro steps:

5. Follow the next step in the debugging instructions, which is:
   > **Start the VS Code debugger**
   > 1\. Choose View > Run or enter Ctrl+Shift+D to switch to debug view.

At this point, a Webpack window appears, and then Excel opens and shows the add-in loaded, and then the Webpack window disappears. Watching the output of the task that gets run, we can see:

```text
Executing task: npm run "start:desktop -- --app excel" 


> excel-custom-functions@0.0.1 start:desktop
> office-addin-debugging start manifest.xml desktop --app excel

Debugging is being started...
App type: desktop
Enabled debugging for add-in f90f0af0-8262-465a-8d46-b5f1a38cc20a.
Starting the dev server... (webpack serve --mode development)
Sideloading the Office Add-in...
Debugging started.
 *  Terminal will be reused by tasks, press any key to close it. 
```

Why is it running dev server?! The instructions have us already running it.

A moment later, the Webpack window disappears and a moment after that, this error pops up in the Debug Console:

```text
Could not read source map for foundation.win32.bundle: Invalid URL
Could not read source map for file:///C:/Program%20Files/Microsoft%20Office/Root/Office16/excel-udf-host.win32.bundle: ENOENT: no such file or directory, open 'c:\Program Files\Microsoft Office\Root\Office16\excel-udf-host.win32.bundle.map'
```

But let's press on.

6. Follow the next step in the debugging instructions, which is *(confusingly, in a third separate section than the previous steps)*:
   > **Start debugging**
   > 1\. In VS Code, open your source code script file (functions.js or functions.ts).
   > 2\. Set a breakpoint in the custom function source code.
   > 3\. In the Excel workbook, enter a formula that uses your custom function.

For our purposes, let's set a breakpoint on the statement in the `add` function. When I do that, it shows as an empty circle, not a red dot, which doesn't bode well. In the Excel window, I click "Got it" on the message pointing to the task pane, then put this in a cell: `=contoso.add(1,2)`

The answer 3 pops up, without hitting the breakpoint.

Here's a summary of the issues and concerns above:

1. Most importantly: What steps can I use to **reliably** debug custom functions in VS Code? Or should I just give up and debug entirely in Edge?
2. Why are there separate `Watch` and `Dev Server` processes that seem to conflict with each other?
   * If they're going to be separate, why does `Dev Server` recompile when `Watch` is going to do that?
3. Why is the `Excel Desktop (Custom Functions)` run task starting `Dev Server` if it's already running?
4. How can I avoid `Watch` going into an endless loop of errors?
5. Why are the debugging steps broken up into sections where the number restarts each time? Why not a cohesive set of "Do this, now do that, now do that" steps?

Thanks.

**Comment by igor-ribeiiro** [comment]
Thanks for the context @tj-aew!
I will try to answer your questions to the best of my knowledge.
`Most importantly: What steps can I use to reliably debug custom functions in VS Code? Or should I just give up and debug entirely in Edge?`
There was another issue recently that I was part of that helped some other customers debug which I put some detailed instructions and also a link to a fix we just made. I recommend you try it out to see if it solves your problem: https://github.com/OfficeDev/Excel-Custom-Functions/issues/336

`Why are there separate Watch and Dev Server processes that seem to conflict with each other?
If they're going to be separate, why does Dev Server recompile when Watch is going to do that?`
There may be some bug there. Can you try avoiding using Watch for now? @Rick-Kirkham, I think we could do some work on this tutorial, I think the `watch` task might be really having a bug in Custom Functions and we could remove it for now.

`Why is the Excel Desktop (Custom Functions) run task starting Dev Server if it's already running?` 
The behavior should be: Try to run, detect that is already running, and not run again. That's what I see when running the dev-server and then running the debugger task. Could you confirm that the dev-server tries to run, but will detect and close the new process?

`How can I avoid Watch going into an endless loop of errors?` 
I would recommend avoiding using the watch for now. I remember there is a bug for hot reloading in CFs which is not working. So I don't think Watch is working right now.

`Why are the debugging steps broken up into sections where the number restarts each time? Why not a cohesive set of "Do this, now do that, now do that" steps?`
I will tag @alison-mk and @Rick-Kirkham for this question.

**Comment by Rick-Kirkham** [comment]
@tj-aew There seem to be 4 different ways of debugging described in the article:

![image](https://user-images.githubusercontent.com/9088056/186529042-c00d8322-86fa-42c4-85a3-00fd4739653b.png)

Can you clarify for me which one(s) are not working? Are any of them working?

**Comment by tj-aew** [comment]
@Rick-Kirkham - Sorry for not being explicit about that, although I did quote the steps. The first one: **Use the VS Code debugger for Excel on Windows**

**Comment by tj-aew** [comment]
@igor-ribeiiro - Thanks. Unfortunately it still doesn't work even if I comment out the `devtoolModuleFilenameTemplate: "webpack:///[resource-path]?[loaders]",` line in `webpack.config.js` (screenshot (A) at end of comment) and avoid using `Watch`. I still get an error in the Debug Console (which may or may not be related) and, more importantly, breakpoints don't work. (Steps I followed below in case I got something wrong.)

> Can you try avoiding using Watch for now?

Will do.

> The behavior should be: Try to run, detect that is already running, and not run again.

I suspect that's what it's doing. It would be great if the debugging page called that out so people knew it would happen. It sets off alarm bells when something seems to start and then end prematurely. (Also, perhaps it could do its "am I already running?" check before showing the various messages about certs and browserlist nag, and show a "Dev server already running elsewhere, not running a second time" message that stays up for 4-5 seconds before the window closes?)

Here are the steps I went through; did I get something wrong?

1. Open the project
   * It auto-runs `Dev Server`
2. Wait until `Dev Server` is not just serving but also finished compiling (takes 30-60 seconds); I only proceed when I see the "webpack 5.52.1 compiled successfully in 43601 ms" message (screenshot (B) at end of comment)
3. Switch to the `Run/Debug` view and run `Excel Desktop (Custom Functions)`
   * At this point, Excel runs
   * A Webpack window opens to run `dev-server`, then closes some seconds later (I can't catch the last message, presumably it's saying `Dev Server` is already running)
   * In Excel, I see the functions are installed
      * Although often at this point, I get an error loading functions and have to try again after clearing `Wef`
4. Type `=contoso.add(1,2)` in a cell. I get the result immediately; my breakpoint is not hit

I've also tried doing `npm run build` as Step 0 above (before even opening VS Code), since I saw in the thread you linked that people were saying to do that (even though `Dev Server` does it).

Separately, whether or not I try to use a custom function, a few seconds after the Excel window loads, I get that error in the Debug Console that I mentioned previously:

> Could not read source map for foundation.win32.bundle: Invalid URL
> Could not read source map for file:///C:/Program%20Files/Microsoft%20Office/Root/Office16/excel-udf-host.win32.bundle: ENOENT: no such file or directory, open 'c:\Program Files\Microsoft Office\Root\Office16\excel-udf-host.win32.bundle.map'

Another couple of notes/questions:
* Why does the `Dev Server` compile take 30-60 seconds when `npm run build` takes about 9?
* Sometimes, when you close Excel, the `Stop Debug` Task doesn't run automatically (I don't know why not), so it leaves a registry entry in `Computer\HKEY_CURRENT_USER\SOFTWARE\Microsoft\Office\16.0\WEF\Developer` (and something in the `Wef` directory?). The debugging page mentions `npm run stop` [at the very end](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-debugging#commands-for-building-and-running-your-add-in). I suggest mentioning it in the main debugging steps, saying it's supposed to run automatically, but that if it doesn't the developer should run it.
* When it does work, what does a dev cycle look like? My guess is:
   1. Start `Dev Server` (project is set to do that automatically)
   2. Iterate:
      1. Run the Excel task
      2. Do some debugging
      3. Exit Excel and make sure Stop Debug runs
      4. Modify your code (`Dev Server` seems to watch for modifications and recompile)
      5. Iterate again
   3. Stop `Dev Server` when done with the dev cycle (or just exit VS Code / switch to another folder/workspace)

Thanks again!

Here are the two screenshots I mentioned above:

(A) Commenting out `devtoolModuleFilenameTemplate`:

![image](https://user-images.githubusercontent.com/101200432/186630101-9aa32d2b-4f44-499c-a1c9-edee10a6da2a.png)

(B) Waiting not just for the `Dev Server` to start serving, but also for it to finish compiling 30-60 seconds later:

![image](https://user-images.githubusercontent.com/101200432/186630368-c42343c9-7888-49ca-abf4-30d024f0b2aa.png)

**Comment by igor-ribeiiro** [comment]
Hello @tj-aew. I just published a new version of Excel Custom Functions (JavaScript only runtime) for TypeScript. Could you test it to see if it works?
Also, I would recommend running the dev-server manually and not using the automated process that is launched at the start. That behavior came with a VSCode update and it was unintentional, so I would recommend avoid relying on it for now.

 `Why does the Dev Server compile take 30-60 seconds when npm run build takes about 9?`
We are using `webpack` internally, so this difference is related to the `webpack` build and `webpack serve`. It is taking longer because it is building and serving the files.


`Here are the steps I went through; did I get something wrong?
1. Try running the dev-server manually. It shouldn't make a difference, but the dev-server autorun was not tested enough, not would be nice not to rely on that.

`Sometimes, when you close Excel, the Stop Debug Task doesn't run automatically`
I don't think it is that bad when it doesn't run. The Stop task ensures that there are no leftovers from the running process, but a lot of times the process clears itself completely, so don't worry if the Stop doesn't run every time.

`When it does work, what does a dev cycle look like? My guess is:`
1. `npm run dev-server`
2. `Run the debug task`
3. Just write some custom functions in Excel (contoso.add), there is no need to open the task pane.
4. See that the debugger attach stops at the breakpoints in the functions files.
5. Stop dev-server when end the process.

**Comment by tj-aew** [comment]
@igor-ribeiiro - Thanks! Sorry for the slow reply, press of work.

Unfortunately, this version doesn't fix the problem. If I follow the steps in your message, the custom functions in the add-in don't load at all:

![image](https://user-images.githubusercontent.com/101200432/188308709-8bddd1d9-3a0e-495b-9c97-53dca1cc1f8e.png)

Here's the error in the `OfficeAddins.log.txt` file:

```text
9/4/2022 11:18:46	Unexpected	Resource	The resource you are trying to use does not exist.		
9/4/2022 11:18:46	Medium	Resource	Resource Download Error Details: SolutionId=1da808d5-6721-47dc-a924-486f8a34bad0 , AppVersion=1.0.0.0		
```

As you can see, the GUID of the error matches the GUID of the workbook that the debugging process created.

I've tried repeatedly, made sure my `Wef` directory was clear, that there were no stray debugging entries in the registry (`HKEY_CURRENT_USER\SOFTWARE\Microsoft\Office\16.0\WEF\Developer`), etc.

If I add the step of running a debug build before doing Debug: Excel Desktop, the custom functions do at least load, but breakpoints don't work. But that seems really odd -- the Dev Server task clearly does a build, so you wouldn't expect to need to do a build as a separate step.

I'm perplexed, how is it that it works at your end and reliably doesn't work at my end? As far as I know, there's nothing odd about my setup. Sorry to ask, but are you starting *from scratch* when testing this? `yo office` and then try to use it?

> Also, I would recommend running the dev-server manually and not using the automated process that is launched at the start.

Agreed, I always turn that off, but I was sticking to the template as provided.

> That behavior came with a VSCode update and it was unintentional, so I would recommend avoid relying on it for now.

It looks pretty intentional -- the project generator creates a `tasks.json` file with Dev Server configured like this:

```json
{
  "label": "Dev Server",
  "type": "npm",
  "script": "dev-server",
  "presentation": {
    "clear": true,
    "panel": "dedicated"
  },
  "problemMatcher": [],
  "runOptions": {
    "runOn": "folderOpen"
  }
},
```

That `"runOn": "folderOpen"` is what makes it run when the folder is open. I suggest removing it (the entire `runOptions` can go since that's the only thing in it).

> > Why does the Dev Server compile take 30-60 seconds when npm run build takes about 9?
> We are using webpack internally, so this difference is related to the webpack build and webpack serve. It is taking longer because it is building and serving the files.

The serving part starts within 3 seconds, it's the build that takes 30-60. But again, an optimized production build from `npm run build` only takes about 8 seconds. `npm run build:dev` takes about 5.5 seconds. It doesn't make any sense for a non-optimized dev build to take 8-10 **times** as long as that when done as part of Dev Server. Can you confirm you see this at your end?

Possibly related: `office-addin-debugging stop manifest.xml` keeps running for 17 seconds after saying it's stopped debugging. This seems to be common behavior for the various `office-xyz` scripts. For instance, `npm run validate` will run, say the manifest is valid, and then linger for 17 seconds before terminating. I wonder if that's related to this startup time issue?

Thanks.

**Comment by igor-ribeiiro** [comment]
Hello @tj-aew, sorry that your problem wasn't solved yet. Thanks for the detailed information you posted, I found some differences in what I'm doing that I will be sharing.

First, about the WEF cache cleaning, I'm not talking about the registry here. You don't need to open the registry at all. Sorry for this misunderstanding.
You need to delete the following directory from your OS: C:\Users\<username>\AppData\Local\Microsoft\Office\16.0\Wef
This is needed when switching to a new custom functions project and is something we are already looking into solving.

`I'm perplexed, how is it that it works at your end and reliably doesn't work at my end? As far as I know, there's nothing odd about my setup. Sorry to ask, but are you starting from scratch when testing this? yo office and then try to use it?`
Yes, I just started from scratch. I really think the WEF cache suggestion I made could be the source cause of this.

`It looks pretty intentional -- the project generator creates a tasks.json file with Dev Server configured like this:`
This code you pasted did another thing in old versions of VSCode. The new version 1.7 of VSCode made the task to run when the project started which wasn't something that was happening before. As it was not tested enough, I would recommend avoid relying on that feature for now.

`The serving part starts within 3 seconds, it's the build that takes 30-60. But again, an optimized production build from npm run build only takes about 8 seconds. npm run build:dev takes about 5.5 seconds. It doesn't make any sense for a non-optimized dev build to take 8-10 times as long as that when done as part of Dev Server. Can you confirm you see this at your end?`
Yes, I see that on my end. I'm not sure what is causing this, but it is nice feedback to have (that the build is taking too long) and we will be aware of that fact.

`office-addin-debugging stop manifest.xml keeps running for 17 seconds after saying it's stopped debugging. This seems to be common behavior for the various office-xyz scripts. For instance, npm run validate will run, say the manifest is valid, and then linger for 17 seconds before terminating. I wonder if that's related to this startup time issue?`
I don't think it is related. From what I know, it is taking a long to make sure the process really stopped. OS process can get trick with a lot of asynchronous thread going around, so it is probably waiting to be extra sure the process is finished.

**Comment by tj-aew** [comment]
@igor-ribeiiro - Thanks. Sorry if I was unclear, I do understand the `Wef` directory thing. When I said *"I've tried repeatedly, made sure my `Wef` directory was clear, that there were no stray debugging entries in the registry (`HKEY_CURRENT_USER\SOFTWARE\Microsoft\Office\16.0\WEF\Developer`), etc."*, I was giving a *list* of things I've tried:

1. made sure my `Wef` directory was clear
2. that there were no stray debugging entries in the registry (`HKEY_CURRENT_USER\SOFTWARE\Microsoft\Office\16.0\WEF\Developer`)
3. etc.

(There's no `Wef` in the registry path anyway.)

For clarity: I am clearing the `Wef` directory in the file system (almost obsessively :-) ) and always starting with a clean one.

> This code you pasted did another thing in old versions of VSCode.

That's...interesting. What did it used to do? I can't immediately imagine another interpretation of `"runOn": "folderOpen"`. :-)

So where do we go from here? My tests are with an empty `Wef`, and it simply doesn't work.

**Comment by igor-ribeiiro** [comment]
@tj-aew I really not entirely sure how we can proceed.
I recommend upgrading all the global packages version you have, such as node to v16.15.0 (the one I'm testing).
I would recommend double-checking your Office/Excel version on the latest.

The only think I can thing of right now is configuration issues. Let me know if this can help solve the issue.

**Comment by tj-aew** [comment]
@igor-ribeiiro - Thanks. Frankly, I **very** much doubt it's a configuration issue. But my Node version is out of date (how time flies). I'll update to the latest LTS (16.17.0) when my IT folks are in the office. Would you do the same? (v16.15.0 is out of date.) If you can't, please let me know ASAP and I'll do v16.15.0 instead, but it's out of date so...

> I recommend upgrading all the global packages version you have

What, other that Node?

Once we're on the same version of Node and if the problem is still occurring, perhaps we could get on a Teams call? I suspect there's some small difference in what we're doing that accounts for it, not configuration. This stuff shouldn't be that delicate.

**Comment by tj-aew** [comment]
@igor-ribeiiro  - Checking back on the above. Are you able to update to the latest LTS, or shall I have my IT department install an out-of-date version of Node.js? Also, please note my question above about global packages.

**Comment by igor-ribeiiro** [comment]
Hello @tj-aew. I would recommend upgrading to 16.15.0 for now. I did some not successful tests on the LTS 16.17.0, which I still have to investigate if it is working or not.
Please, test on 16.15.0 and let me know the results.

**Comment by bencassie** [comment]
I am experiencing the same issue as @tj-aew.  

Has the solution moved on in the last couple months at all?

**Comment by tj-aew** [comment]
FWIW @bencassie I gave up on Microsoft fixing this and just debug using the Edge devtools. It was taking too much of my time to keep going back and forth with no progress. (Obviously it has zero to do with the Node.js version.)

**Comment by Rick-Kirkham** [comment]
@tj-aew @bencassie Although we aren't able to reproduce the issues with the debugging technique described in that first section of the article, it seems that the technique is causing a lot of problems. Since there are other techniques described in the same article that work more reliably, we have simply removed that section from the article. I will close this issue now. Thanks for your interest in Office Add-ins.

**Comment by tj-aew** [comment]
@Rick-Kirkham - So now the add-in generator creates command in the `package.json` scripts that are undocumented (and don't work)? Or is the generator going to stop creating those, making debugging in VS Code unsupported? There are a number of advantages to debugging in VS Code, rather than in the browser, if Microsoft can make it work.

**Comment by Rick-Kirkham** [comment]
@tj-aew As I understand it, the script _does_ work. We can't reproduce the problems. It is now undocumented, however. 

At some point, the generator will be revised to stop producing it, but I don't know when that will happen. We are planning bigger changes in the tooling, across the M365 landscape, that will tend to make this issue moot. The current Yeoman generator may go away or be merged into another tool. 

As for debugging in the browser, Microsoft has been moving in that direction for some time. The debugger in Visual Studio (not VS Code), hasn't worked with add-ins for some time. Our article on debugging in Visual Studio basically describes how to debug in the browser. Similarly, to debug on a Mac, we tell readers to use the Safari tools. I don't see this trend changing anytime soon.  It is costly for Microsoft to maintain redundant debugging systems.
